File: leetcode/p00001/p00001.py
from typing import List

class Solution:

    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = {}

        for idx, val in enumerate(nums):
            remaining = target - val
            if remaining in seen:
                return [seen[remaining], idx]

            seen[val] = idx

        return None

if __name__ == '__main__':
    s = Solution()

    assert s.twoSum([2, 7, 11, 15], 9) == [0, 1]
    assert s.twoSum([3, 2, 4], 6) == [1, 2]
    assert s.twoSum([3, 3], 6) == [0, 1]

File: leetcode/p00001/p00001.ts
import assert from "assert";

// Time Complexity: O(n)
// Space Complexity: O(n)
function twoSum(nums: number[], target: number): number[] {
    const numMap = new Map<number, number>();
    
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        
        if (numMap.has(complement)) {
            return [numMap.get(complement)!, i];
        }
        
        numMap.set(nums[i], i);
    }
    
    return [];
}


if (require.main === module) {
    const result1 = twoSum([2, 7, 11, 15], 9);
    assert.deepStrictEqual(result1, [0, 1], "Test 1 failed");

    const result2 = twoSum([3, 2, 4], 6);
    assert.deepStrictEqual(result2, [1, 2], "Test 2 failed");
    
    const result3 = twoSum([3, 3], 6);
    assert.deepStrictEqual(result3, [0, 1], "Test 3 failed");
}